### PR TITLE
Simplified Builder for HoneyClient

### PR DESCRIPTION
**Pull request by MarsianMan**
- Added HoneyClientBuilder to ease HoneyClient instance creation. See README in io.honeycomb.libhoney.builders package for usage. 
- Added DefaultDebugResponseObserver to provide debug ResponseObserver for users who directly use libhoney. 
  - HoneyClientBuilder will add DefaultDebugResponseObserver to HoneyClient ResponseObserver list when `debug(true)` is called on the builder. 
  - Provide explanatory messages for scenarios when Honeycomb API server cannot be reached.
- Bumped version to snapshot.

### CODE REVIEW

File: libhoney/src/main/java/io/honeycomb/libhoney/builders/README.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    public Example(){
+        client = new HoneyClientBuilder()
+                    .dataSet("test-dataset")
+                    .writeKey("WRITE_KEY")
+                    .transport(new MockTransport())
+                    .build();
+    }
+}
+```
```
Comment (paulosman) [inline libhoney/src/main/java/io/honeycomb/libhoney/builders/README.md:96]: @nathanleclaire can you take a look at the examples in this README and provide any feedback?

File: libhoney/src/test/java/io/honeycomb/libhoney/End2EndTest.java [hunk 2]
```diff
@@ -27,6 +27,10 @@
 import static io.honeycomb.libhoney.TransportOptions.DEFAULT_BATCH_SIZE;
 import static org.assertj.core.api.Assertions.assertThat;
 
+/**
+ * This class defines the test spec of the client. Any upate to this should be reflected in {@link End2EndBuilderTest}
+ *
```
Comment (paulosman) [inline libhoney/src/test/java/io/honeycomb/libhoney/End2EndTest.java]: nit: extra blank line

File: libhoney/src/main/java/io/honeycomb/libhoney/DefaultDebugResponseObserver.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+public class DefaultDebugResponseObserver implements ResponseObserver {
+    private static final Logger LOG = LoggerFactory.getLogger(DefaultDebugResponseObserver.class);
+
+    protected static final String ERROR_TEMPLATE_401 = "Server responded with a 401 HTTP error code to a batch request." +
+        " This is likely caused by using an incorrect 'Team Write Key'. Check https://ui.honeycomb.io/account to verify your " +
+        "team write key. Rejected event: {}";
+    protected static final String ERROR_TEMPLATE_UNKNOWN_HOST_EXCEPTION = "Could not reach determine destination server. " +
+        "This could be due to 1) invalid apiHost 2) Could not resolve apiHost (firewall? dns?) 3) problem with proxy configuration. " +
+        "Unknown event: {}";
```
Comment (paulosman) [inline libhoney/src/main/java/io/honeycomb/libhoney/DefaultDebugResponseObserver.java:18]: Not sure the word "determine" belongs here? Could we rewrite as "Could not reach destination server...". I think that's clearer.

File: libhoney/src/main/java/io/honeycomb/libhoney/DefaultDebugResponseObserver.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Override
+    public void onClientRejected(final ClientRejected clientRejected) {
+        LOG.debug("Event rejected on the client side: {}", clientRejected);
+    }
+
+    @Override
+    public void onUnknown(final Unknown unknown) {
+        if(unknown.getException()!=null && UnknownHostException.class.isAssignableFrom(unknown.getException().getClass())){
```
Comment (paulosman) [inline libhoney/src/main/java/io/honeycomb/libhoney/DefaultDebugResponseObserver.java]: I don't think we've specified a style guide, but let's add spaces around binary operators:

```
if (unknown.getException() != bill && UnknownHostException.class.isAssignableFrom(unknown.getException().getClass())) {
```

Might also be worth adding a line break there.

File: libhoney/src/main/java/io/honeycomb/libhoney/builders/README.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+public class Example{
+    HoneyClient client;
+
+    public Example(){
+        client = new HoneyClientBuilder()
+                    .dataSet("test-dataset")
+                    .writeKey("WRITE_KEY")
+                    .addProxyNoCredential("https://myproxy.example.com")
+                    .build();
```
Comment (paulosman) [inline libhoney/src/main/java/io/honeycomb/libhoney/builders/README.md:70]: :raised_hands: SO much nicer.

File: libhoney/src/test/java/io/honeycomb/libhoney/End2EndBuilderTest.java [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+    @Rule
+    public WireMockRule wireMock = new WireMockRule(8089);
+
+    private HoneyClient honeyClient;
+    private BlockingQueue<Response> notifyQueue;
+
+    @Before
+    public void setup() {
+        stubServer(200, "[" +
+            "  {\"status\": 202}" +
```
Comment (paulosman) [inline libhoney/src/test/java/io/honeycomb/libhoney/End2EndBuilderTest.java:67]: Why `200` as the status and `202` in the response body? If that's intentional, it's probably worth a comment.

Also, I know it's just a test, but maybe we can replace these with constants? At least for the status, maybe a bit much for the body.

File: libhoney/src/test/java/io/honeycomb/libhoney/End2EndBuilderTest.java [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+import static io.honeycomb.libhoney.TransportOptions.builder;
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * This ensures HoneyClientBuilder created clients match expected results as defined by {@link End2EndTest}.
+ */
+public class End2EndBuilderTest {
+
+    @Rule
+    public WireMockRule wireMock = new WireMockRule(8089);
```
Comment (paulosman) [inline libhoney/src/test/java/io/honeycomb/libhoney/End2EndBuilderTest.java]: Let's replace the port number with a constant for clarity

File: libhoney/src/test/java/io/honeycomb/libhoney/End2EndBuilderTest.java [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+
+    // this client will not timeout batches
+    private void createDefaultClient() throws URISyntaxException {
+        honeyClient = createBuilder().build();
+        notifyQueue = createObserverQueue();
+    }
+
+
+    private HoneyClientBuilder createBuilder() throws URISyntaxException {
+        return new HoneyClientBuilder().writeKey("testWriteKey").dataSet("testDataSet").apiHost("http://localhost:8089");
```
Comment (paulosman) [inline libhoney/src/test/java/io/honeycomb/libhoney/End2EndBuilderTest.java]: Would rather `"http://localhost:8089"` be something like `LOCAL_TEST_URL` or similar.

File: libhoney/src/main/java/io/honeycomb/libhoney/builders/README.md (hunk 5 above)
Comment (paulosman) [reply libhoney/src/main/java/io/honeycomb/libhoney/builders/README.md:70]: Can we provide an example with credentials as well? That's pretty common in my experience.

**Review by paulosman — CHANGES_REQUESTED** [review]
Looks great, thank you! If you address the comments and @nathanleclaire gets a chance to review the README snippets, I'll :+1:

File: libhoney/src/test/java/io/honeycomb/libhoney/End2EndTest.java (hunk 2 above)
Comment (MarsianMan) [reply libhoney/src/test/java/io/honeycomb/libhoney/End2EndTest.java]: Removed.

File: libhoney/src/main/java/io/honeycomb/libhoney/DefaultDebugResponseObserver.java (hunk 3 above)
Comment (MarsianMan) [reply libhoney/src/main/java/io/honeycomb/libhoney/DefaultDebugResponseObserver.java:18]: Agreed and removed.

File: libhoney/src/main/java/io/honeycomb/libhoney/DefaultDebugResponseObserver.java (hunk 4 above)
Comment (MarsianMan) [reply libhoney/src/main/java/io/honeycomb/libhoney/DefaultDebugResponseObserver.java]: Good call on the operator spacing. I did not add a new line as it was on the edge of too long but a simple statement. If you feel strongly about it, I can add a line.

File: libhoney/src/test/java/io/honeycomb/libhoney/End2EndBuilderTest.java (hunk 6 above)
Comment (MarsianMan) [reply libhoney/src/test/java/io/honeycomb/libhoney/End2EndBuilderTest.java:67]: This might seem like a "WTF" on first look, but that is because they are status flags for different purposes. 

202 is defined by Honeycomb API on [batching](https://docs.honeycomb.io/api/events/#batched-events) as "enqueued for processing" (application level flag). The Honeycomb framework uses HTTP transport to send events and the 200 comes from the HTTP protocol response (transport level flag).

I have added comments that hopefully help explain the issue.

File: libhoney/src/main/java/io/honeycomb/libhoney/DefaultDebugResponseObserver.java (hunk 4 above)
Comment (MarsianMan) [reply libhoney/src/main/java/io/honeycomb/libhoney/DefaultDebugResponseObserver.java]: BTW, found the `!= bill` hilarious in your quote. 😄

File: libhoney/src/main/java/io/honeycomb/libhoney/builders/README.md (hunk 5 above)
Comment (MarsianMan) [reply libhoney/src/main/java/io/honeycomb/libhoney/builders/README.md:70]: Added below and unified the naming of the functions.

File: libhoney/src/test/java/io/honeycomb/libhoney/End2EndBuilderTest.java (hunk 7 above)
Comment (MarsianMan) [reply libhoney/src/test/java/io/honeycomb/libhoney/End2EndBuilderTest.java]: Done

File: libhoney/src/test/java/io/honeycomb/libhoney/End2EndBuilderTest.java (hunk 8 above)
Comment (MarsianMan) [reply libhoney/src/test/java/io/honeycomb/libhoney/End2EndBuilderTest.java]: Done.

File: libhoney/src/test/java/io/honeycomb/libhoney/End2EndBuilderTest.java (hunk 6 above)
Comment (paulosman) [reply libhoney/src/test/java/io/honeycomb/libhoney/End2EndBuilderTest.java:67]: Aha, TIL! Thank you, the comment is helpful.

File: libhoney/src/main/java/io/honeycomb/libhoney/DefaultDebugResponseObserver.java (hunk 4 above)
Comment (paulosman) [reply libhoney/src/main/java/io/honeycomb/libhoney/DefaultDebugResponseObserver.java]: Haha. That must have been autocorrected. I like it :) 

No worries about the newline. More concerned about the spacing.

**Comment by nathanleclaire** [comment]
Hey @MarsianMan , thanks for the PR!

A couple of comments on my end:

1. Thank you for adding the `DefaultDebugResponseObserver`! We've been needing that.
2. Can you comment a little on the motivations for the change? i.e., what will be better using the new builder instead of `LibHoney.create(LibHoney.options().setDataset(DATASET).setWriteKey(WRITE_KEY).build())` where a lot of the methods in the `TransportOptions.Builder` mirror the builder here?

**Comment by MarsianMan** [comment]
> 2. Can you comment a little on the motivations for the change? i.e., what will be better using the new builder instead of `LibHoney.create(LibHoney.options().setDataset(DATASET).setWriteKey(WRITE_KEY).build())` where a lot of the methods in the `TransportOptions.Builder` mirror the builder here?

A problem for Beeline users is that they have to manually create the HoneyClient to set options. In addition, the configuration of Beeline itself and the configuration of Beeline's underlying HTTP client framework are configured in two different spots. This experience was reported as cumbersome.

This new builder (paired with https://github.com/honeycombio/beeline-java/pull/43) improves the abstraction of the Beeline and HoneyClient. Users will not have to know libhoney internals in order to create either the HoneyClient or Beeline. In addition, this facilitates passing through configuration from the Beeline creation to the underlying HoneyClient instance with _minimal logic duplication_ (i.e. instance creation details).  Meaning, BeelineBuilder internally calls HoneyClientBuilder and does not duplicate HoneyClientBuilder's logic.

I think the similar naming of Transport.Options and Options does not help either.

Everyone is switching to Spring Boot for the simplicity in use and configuration, after all.  😃 The Spring Boot starter hides some of this complexity. _Even that_ still does not allow full configuration (e.g. proxy) without manually creating the HoneyClient instance. Moving to Spring Boot is not an option/desirable for some users.

### tl;dr Let's look at an example
Let's compare setting proxy
*Current API*
```java
// From https://docs.honeycomb.io/getting-data-in/java/sdk/#working-with-a-proxy
public class Initialization {
    private HoneyClient honeyClient;

    public Initialization() {
        honeyClient = create(
            options()
                .setWriteKey("myTeamWriteKey")
                .setDataset("Cluster Dataset")
                .setSampleRate(2)
                .build(), // Options builder
            transportOptions() // second builder needed for confusingly named Transport.Options
                .setProxy(HttpHost.create("https://myproxy.example.com"))
                .build() 
        ); // create is a factory that accepts *2 builders* because we STILL do not have a client
    }
}
```
*Proposed API*
```java
public class Initialization {
    private HoneyClient honeyClient;

    public Initialization() {
        honeyClient = new HoneyClientBuilder().writeKey("myTeamWriteKey")
               .dataSet("Cluster Dataset")
               .sampleRate(2)
              .addProxy("https://myproxy.example.com")
              .build() // client does not have to care about implementation of configuration, just set the configuration and don't worry about it.
        );
    }
}
```
Doing the same task is even more complicated from the Beeline creation perspective!
